File: examples/middleware/provider-sdks/certification-native.mjs
/** Frozen operation journeys on official SDK clients and the real loopback runtime. */
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { readFile } from 'node:fs/promises';
import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';
import { withCavemanOpenAI, withCavemanOpenAITools } from '../../../packages/middleware/typescript/dist/openai.js';
import { withCavemanAnthropic } from '../../../packages/middleware/typescript/dist/anthropic.js';
import { SOURCE, FACT, api, args, definitions, instrument, providerFixture, scope, textOf, waitFor } from './native-fixture.mjs';
import { beginNativeCertification, emitJourney } from './certification-evidence.mjs';

const digest = value => createHash('sha256').update(typeof value === 'string' ? value : JSON.stringify(value)).digest('hex');
const sourceHash = digest(SOURCE);
const protocolOf = cell => cell.protocol === 'openai-chat-completions' ? 'openai-chat' : cell.protocol;
const eventName = event => event.type ?? event.object ?? event.constructor.name;
const collect = async stream => { const output = []; for await (const event of stream) output.push(event); return output; };
const resultText = (value, protocol) => protocol === 'openai-responses' ? value.output_text ?? value.output[0].content[0].text : textOf(value, protocol);
const results = (body, protocol) => Object.fromEntries(protocol === 'openai-chat' ? (body.messages ?? []).filter(item => item.role === 'tool').map(item => [item.tool_call_id, item.content]) :
  protocol === 'openai-responses' ? (body.input ?? []).filter(item => item.type === 'function_call_output').map(item => [item.call_id, item.output]) :
  (body.messages ?? []).flatMap(item => Array.isArray(item.content) ? item.content : []).filter(item => item.type === 'tool_result').map(item => [item.tool_use_id, item.content]));
function nativeClient(provider, protocol) {
  return protocol.startsWith('openai') ? new OpenAI({ apiKey: 'local-fixture', baseURL: provider.url + '/v1', maxRetries: 0, timeout: 3000 }) :
    new Anthropic({ apiKey: 'local-fixture', baseURL: provider.url, maxRetries: 0, timeout: 3000 });
}
const wrap = (client, protocol, options) => protocol.startsWith('openai') ? withCavemanOpenAI(client, options) : withCavemanAnthropic(client, options);
function reportEvidence(observed, count, mode) {
  assert.equal(observed.reports.length, count, 'exactly one report per wrapped physical provider attempt');
  assert.equal(observed.runtime.lastReport, observed.reports.at(-1));
  for (const report of observed.reports) {
    assert.ok(Object.isFrozen(report) && Object.isFrozen(report.transform_ids));
    const replacements = observed.plans.find(plan => plan.result.request?.logical_call_id === report.logical_call_id && plan.result.request?.attempt_id === report.attempt_id)?.result.replacements ?? [];
    assert.equal(report.status, mode === 'off' ? 'disabled' : replacements.length ? replacements.every(item => item.reused) ? 'reused' : 'applied' : 'skipped');
    assert.equal(report.replacement_count, replacements.length);
    assert.deepEqual(report.transform_ids, [...new Set(replacements.map(item => item.transform_id))].sort());
    assert.ok(!JSON.stringify(report).includes(SOURCE));
  }
  if (mode === 'off') { assert.equal(observed.plans.length, 0); assert.equal(observed.receipts.length, 0); }
  return { count, statuses: observed.reports.map(report => report.status), replacement_counts: observed.reports.map(report => report.replacement_count), metadata_only: true };
}
function appendResponse(messages, reply, protocol) {
  if (protocol === 'openai-chat') {
    messages.push(reply.choices[0].message);
    return (reply.choices[0].message.tool_calls ?? []).map(call => ({ id: call.id, name: call.function.name, input: JSON.parse(call.function.arguments) }));
  }
  if (protocol === 'openai-responses') {
    messages.push(...reply.output);
    return reply.output.filter(item => item.type === 'function_call').map(call => ({ id: call.call_id, name: call.name, input: JSON.parse(call.arguments) }));
  }
  messages.push({ role: 'assistant', content: reply.content });
  return reply.content.filter(item => item.type === 'tool_use').map(call => ({ id: call.id, name: call.name, input: call.input }));
}
function appendResult(messages, protocol, id, value) {
  const content = typeof value === 'string' ? value : JSON.stringify(value);
  messages.push(protocol === 'openai-chat' ? { role: 'tool', tool_call_id: id, content } :
    protocol === 'openai-responses' ? { type: 'function_call_output', call_id: id, output: content } :
    { role: 'user', content: [{ type: 'tool_result', tool_use_id: id, content }] });
}
async function sourceHistory(original, protocol, executions) {
  const messages = [{ role: 'user', content: 'Read source and recover row 70' }];
  const reply = await api(original, protocol).create({ ...args(protocol, 'loop', messages), tools: definitions(protocol) });
  const calls = appendResponse(messages, reply, protocol);
  assert.deepEqual(calls, [{ id: 'read-1', name: 'read_logs', input: {} }]);
  const readLogs = async input => { executions.push(input); return SOURCE; };
  appendResult(messages, protocol, calls[0].id, await readLogs(calls[0].input));
  return { messages, responseId: reply.id };
}

async function targetMethod(cell, client, provider, messages, responseId) {
  const protocol = protocolOf(cell), resource = api(client, protocol), params = args(protocol, 'helpers', messages);
  let native, value, events = [];
  if (['create', 'messages.create'].includes(cell.method)) {
    native = await resource.create(params); value = resultText(native, protocol);
  } else if (['create.stream', 'messages.create.stream'].includes(cell.method)) {
    native = await resource.create({ ...params, stream: true });
    const chunks = await collect(native); events = chunks.map(eventName); assert.ok(events.length > 2);
    value = chunks.map(event => protocol === 'openai-chat' ? event.choices[0]?.delta.content ?? '' :
      protocol === 'openai-responses' ? event.type === 'response.output_text.delta' ? event.delta : '' :
      event.type === 'content_block_delta' && event.delta.type === 'text_delta' ? event.delta.text : '').join('');
  } else if (cell.method === 'asResponse') {
    native = await resource.create(params).asResponse(); assert.equal(native.status, 200); assert.equal(native.bodyUsed, false);
    value = resultText(await native.json(), protocol);
  } else if (['withResponse', 'messages.raw_response'].includes(cell.method)) {
    const result = await resource.create(params).withResponse(); assert.equal(result.response.status, 200);
    native = result.data; value = resultText(native, protocol); assert.ok(result.request_id);
  } else if (cell.method === 'parse') {
    const schema = { type: 'object', properties: { answer: { type: 'integer' } }, required: ['answer'], additionalProperties: false };
    native = await resource.parse({ ...params, model: 'parse', ...(protocol === 'openai-chat' ?
      { response_format: { type: 'json_schema', json_schema: { name: 'answer', schema, strict: true } } } :
      { text: { format: { type: 'json_schema', name: 'answer', schema, strict: true } } }) });
    value = protocol === 'openai-chat' ? native.choices[0].message.parsed : native.output_parsed;
    assert.deepEqual(value, { answer: 42 });
  } else if (cell.method === 'messages.stream') {
    native = resource.stream(params); events = (await collect(native)).map(eventName);
    value = resultText(await native.finalMessage(), protocol); assert.ok(events.length > 2);
  } else if (cell.method === 'cancel') {
    native = await resource.create({ ...params, stream: true });
    const iterator = native[Symbol.asyncIterator](), first = await iterator.next();
    assert.equal(first.done, false); events.push(eventName(first.value)); native.controller.abort();
    assert.equal(provider.released, false); const closure = provider.closures.at(-1).promise;
    assert.equal(await waitFor(closure), true); assert.equal((await closure).beforeRelease, true);
    value = 'cancelled_before_fixture_eof';
  } else if (cell.method === 'server_history_reference') {
    for (const reference of [{ previous_response_id: responseId }, { conversation: 'conv-fixture' }, { conversation: { id: 'conv-fixture' } }]) {
      native = await resource.create({ model: 'helpers', input: [messages.at(-1)], ...reference });
      assert.equal(resultText(native, protocol), 'native'); assert.deepEqual(provider.calls.at(-1).body.input, [messages.at(-1)]);
      for (const [key, entry] of Object.entries(reference)) assert.deepEqual(provider.calls.at(-1).body[key], entry);
    }
    value = 'native';
  } else if (cell.method === 'unrelated_endpoint_passthrough') {
    native = await client.embeddings.create({ model: 'fixture-embedding', input: SOURCE, encoding_format: 'float' });
    value = native.data[0].embedding; assert.deepEqual(value, [1, 2]);
  } else if (cell.method === 'countTokens.passthrough') {
    native = await client.messages.countTokens({ model: 'helpers', messages });
    value = { input_tokens: native.input_tokens }; assert.deepEqual(value, { input_tokens: 7 });
  } else throw new Error(`No native method for ${cell.id}`);
  if (!['parse', 'cancel', 'unrelated_endpoint_passthrough', 'countTokens.passthrough'].includes(cell.method)) assert.equal(value, 'native');
  return { native_type: native.constructor.name, value, events };
}

async function modelCase(cell, endpoint, mode) {
  const protocol = protocolOf(cell), provider = await providerFixture(protocol, { pause: cell.method === 'cancel' ? 'first' : undefined });
  const observed = instrument(endpoint, mode, 3000);
  try {
    if (mode === 'compress') await observed.runtime.ready();
    const original = nativeClient(provider, protocol), client = wrap(original, protocol, { runtime: observed.runtime, scope: scope(digest(cell.id + mode)), fetch });
    const executions = [], { messages, responseId } = await sourceHistory(original, protocol, executions), before = structuredClone(messages);
    const output = await targetMethod(cell, client, provider, messages, responseId);
    assert.deepEqual(messages, before); assert.deepEqual(executions, [{}]); assert.deepEqual(provider.errors, []);
    assert.ok(observed.plans.every(plan => plan.result.replacements.length === 0 && plan.options.binding === null));
    assert.ok(JSON.stringify(provider.calls.at(-1).body).includes(JSON.stringify(SOURCE).slice(1, -1)));
    const noOptimizer = ['server_history_reference', 'unrelated_endpoint_passthrough', 'countTokens.passthrough'].includes(cell.method) || mode === 'off';
    assert.equal(observed.plans.length, noOptimizer ? 0 : 1);
    if (mode === 'outage' && observed.plans.length) assert.equal(observed.plans.at(-1).result.reason, 'runtime_unavailable');
    return { output, reports: reportEvidence(observed, provider.calls.length - 1, mode), source_executions: executions.length, provider_calls: provider.calls.length, optimizer_requests: observed.plans.length,
      recovery_requests: 0, replacements: 0, request_sha256: digest(provider.calls.at(-1).raw),
      response_sha256: cell.recovery === 'not_applicable' ? digest(provider.responses.at(-1).raw) : null,
      history_sha256: digest(before), original_provider_text: true, native_socket_closed_before_eof: cell.method === 'cancel' };
  } finally { observed.runtime.close(); await provider.close(); }
}

async function loopCase(cell, endpoint, mode) {
  const protocol = protocolOf(cell), provider = await providerFixture(protocol), observed = instrument(endpoint, mode, 3000);
  try {
    if (mode === 'compress') await observed.runtime.ready();
    const original = nativeClient(provider, protocol), options = { runtime: observed.runtime, scope: scope(digest(cell.id + mode)), fetch };
    const sourceCalls = [], eventTypes = [], messages = [{ role: 'user', content: 'Read source and recover row 70' }], before = structuredClone(messages);
    const readLogs = async input => { sourceCalls.push(input); return SOURCE; };
    let native, final, stored, executionOwner;
    if (protocol === 'openai-chat') {
      const client = withCavemanOpenAI(original, options);
      const runner = client.chat.completions.runTools({ model: 'loop', messages, stream: cell.streaming,
        tools: [{ type: 'function', function: { ...definitions(protocol)[0].function, parse: JSON.parse, function: readLogs } }] }, { maxChatCompletions: 5 });
      for (const event of ['message', 'functionCall', 'functionCallResult', 'finalContent']) runner.on(event, () => eventTypes.push(event));
      final = await runner.finalContent(); native = await runner.finalChatCompletion();
      stored = results({ messages: runner.messages }, protocol); executionOwner = 'native_tool_runner'; assert.deepEqual(messages, before);
    } else if (protocol === 'openai-responses') {
      assert.equal(typeof original.responses.runTools, 'undefined');
      const bundle = withCavemanOpenAITools(original, { ...options, protocol, tools: definitions(protocol), functions: { read_logs: readLogs } });
      for (let step = 0; step < 5; step++) {
        const params = { ...args(protocol, 'loop', messages), tools: bundle.tools };
        if (cell.streaming) {
          const stream = bundle.client.responses.stream(params); eventTypes.push((await collect(stream)).map(eventName)); native = await stream.finalResponse();
        } else native = await bundle.client.responses.create(params);
        const calls = appendResponse(messages, native, protocol);
        if (!calls.length) break;
        for (const call of calls) appendResult(messages, protocol, call.id, await bundle.functions[call.name](call.input));
      }
      final = resultText(native, protocol); stored = results({ input: messages }, protocol); executionOwner = 'application';
    } else {
      const client = withCavemanAnthropic(original, options), runner = client.beta.messages.toolRunner({ model: 'loop', max_tokens: 100, max_iterations: 5,
        messages, stream: cell.streaming, tools: [{ ...definitions(protocol)[0], parse: input => input, run: readLogs }] });
      if (cell.streaming) {
        for await (const stream of runner) { eventTypes.push((await collect(stream)).map(eventName)); native = await stream.finalMessage(); }
        assert.equal(resultText(await runner.done(), protocol), FACT);
      } else native = await runner;
      final = resultText(native, protocol); stored = results(runner.params, protocol); executionOwner = 'native_tool_runner'; assert.deepEqual(messages, before);
    }
    assert.equal(final, FACT); assert.deepEqual(sourceCalls, [{}]); assert.equal(stored['read-1'], SOURCE); assert.deepEqual(provider.errors, []);
    assert.equal(provider.calls.length, mode === 'compress' ? 3 : 2);
    const projected = results(provider.calls[1].body, protocol)['read-1'], recovered = results(provider.calls.at(-1).body, protocol)['recover-1'];
    assert.equal(projected !== SOURCE, mode === 'compress');
    if (mode === 'compress') {
      assert.ok(!projected.includes(FACT) && projected.includes('cmw_')); assert.equal(JSON.parse(recovered).text, SOURCE);
      assert.ok(observed.plans.some(plan => plan.result.replacements.length));
    } else {
      assert.equal(recovered, undefined); assert.ok(observed.plans.every(plan => plan.result.replacements.length === 0));
      assert.equal(observed.plans.length, mode === 'off' ? 0 : 2);
      if (mode === 'outage') assert.ok(observed.plans.some(plan => plan.result.reason === 'runtime_unavailable'));
    }
    return { reports: reportEvidence(observed, provider.calls.length, mode), execution_owner: executionOwner, upstream_scheduler_available: protocol !== 'openai-responses', native_type: native.constructor.name, value: final,
      events: eventTypes, source_executions: sourceCalls.length, provider_calls: provider.calls.length, optimizer_requests: observed.plans.length,
      recovery_requests: Number(recovered !== undefined), recovered_sha256: recovered === undefined ? null : digest(JSON.parse(recovered).text),
      stored_source_sha256: digest(stored['read-1']), transformed_provider_requests: provider.calls.filter(call => results(call.body, protocol)['read-1'] !== undefined && results(call.body, protocol)['read-1'] !== SOURCE).length };
  } finally { observed.runtime.close(); await provider.close(); }
}

export async function certifyCells(t, family, endpoint) {
  const response = await fetch(`${endpoint}/caveman/v1/middleware/capabilities`); assert.equal(response.status, 200);
  await beginNativeCertification(t, family, await response.json());
  const fixture = JSON.parse(await readFile(new URL('./certification-cells.json', import.meta.url), 'utf8'));
  if (process.env.CAVEMAN_MIDDLEWARE_PACKAGED_TEST !== '1') {
    const { requiredCells } = await import('../../../packages/middleware/conformance/support/catalog.mjs');
    assert.deepEqual(fixture.cells, requiredCells().filter(cell => ['F01', 'F02'].includes(cell.family) && cell.language === 'typescript'));
  }
  const cells = fixture.cells.filter(cell => cell.family === family);
  assert.equal(cells.length, family === 'F01' ? 19 : 8);
  for (const cell of cells) {
    const rows = {};
    for (const mode of ['compress', 'off', 'outage']) rows[mode] = await (cell.recovery === 'native_executor' ? loopCase(cell, endpoint, mode) : modelCase(cell, endpoint, mode));
    const { compress: current, off, outage: unavailable } = rows;
    let journey;
    if (cell.recovery === 'native_executor') {
      assert.equal(current.value, off.value); assert.equal(current.value, unavailable.value);
      journey = {
        native_application: { outcome: 'observed', execution_owner: current.execution_owner, upstream_scheduler_available: current.upstream_scheduler_available, method: cell.method, execution: cell.execution },
        real_tool_result: { outcome: 'observed', executor: 'read_logs', executions: current.source_executions, utf8_bytes: Buffer.byteLength(SOURCE), sha256: sourceHash },
        transformed_provider_request: { outcome: 'observed', provider_requests_with_projection: current.transformed_provider_requests, omitted_fact_absent: true, stored_source_sha256: current.stored_source_sha256 },
        omitted_fact_requested: { outcome: 'observed', requested_fact: FACT, native_recovery_function: 'caveman_retrieve', recovery_requests: current.recovery_requests },
        host_executes_exact_recovery: { outcome: 'observed', execution_owner: current.execution_owner, recovered_sha256: current.recovered_sha256, source_sha256: sourceHash },
        native_result_history_events_and_call_count: { outcome: 'observed', native_reports: current.reports, native_type: current.native_type, final_value: current.value, event_order: current.events, provider_calls: current.provider_calls, stored_source_sha256: current.stored_source_sha256 },
        off_baseline: { outcome: 'observed', native_reports: off.reports, final_value: off.value, provider_calls: off.provider_calls, optimizer_requests: off.optimizer_requests, recovery_requests: off.recovery_requests, stored_source_sha256: off.stored_source_sha256 },
        optimizer_unavailable: { outcome: 'observed', native_reports: unavailable.reports, final_value: unavailable.value, provider_calls: unavailable.provider_calls, optimizer_requests: unavailable.optimizer_requests, recovery_requests: unavailable.recovery_requests, stored_source_sha256: unavailable.stored_source_sha256 },
      };
    } else {
      assert.deepEqual(current.output, off.output); assert.deepEqual(current.output, unavailable.output);
      for (const baseline of [off, unavailable]) { assert.equal(current.request_sha256, baseline.request_sha256); assert.equal(current.provider_calls, baseline.provider_calls); }
      const passthrough = cell.recovery === 'not_applicable';
      const noRecovery = { outcome: 'recovery_free', reason: passthrough ? 'non-generation endpoint preserves the native exchange' : 'native model-only operation has no registered recovery executor', recovery_requests: 0, replacements: 0,
        ...(passthrough ? { applicability: 'endpoint_passthrough', optimizer_requests: 0, request_sha256: current.request_sha256, response_sha256: current.response_sha256, provider_calls: current.provider_calls } : { original_provider_text: current.original_provider_text, source_sha256: sourceHash }) };
      if (passthrough) for (const baseline of [off, unavailable]) assert.equal(current.response_sha256, baseline.response_sha256);
      const baseline = value => ({ outcome: 'observed', native_reports: value.reports, native_output: value.output, ...(passthrough ? { applicability: 'endpoint_passthrough' } : {}),
        ...Object.fromEntries(['request_sha256', 'response_sha256', 'provider_calls', 'optimizer_requests', 'recovery_requests', 'replacements'].map(key => [key, value[key]])) });
      journey = {
        native_application: { outcome: 'observed', method: cell.method, execution: cell.execution, official_sdk_client: true },
        real_tool_result: passthrough ? noRecovery : { outcome: 'observed', executor: 'read_logs', execution_owner: 'application', executions: current.source_executions, sha256: sourceHash, utf8_bytes: Buffer.byteLength(SOURCE) },
        transformed_provider_request: noRecovery, omitted_fact_requested: noRecovery, host_executes_exact_recovery: noRecovery,
        native_result_history_events_and_call_count: { outcome: 'observed', native_reports: current.reports, ...current.output, provider_calls: current.provider_calls, original_history_sha256: current.history_sha256, native_socket_closed_before_eof: current.native_socket_closed_before_eof },
        off_baseline: baseline(off), optimizer_unavailable: baseline(unavailable),
      };
    }
    emitJourney(t, cell, journey);
  }
}

File: examples/middleware/litellm/_certification_native.py
"""Exact pinned LiteLLM SDK, Router and Proxy operation journeys."""
from __future__ import annotations
import asyncio
import copy
import hashlib
import json
import os
import re
from dataclasses import asdict
from pathlib import Path

import litellm
from caveman_cloud.middleware import MiddlewareRuntime, Scope
from caveman_middleware.litellm import CavemanLiteLLM
from _provider import Provider, SOURCE, FACT, history, definitions, results

TEST_FILE = 'examples/middleware/litellm/test_native.py'
CELLS = json.loads(Path(__file__).with_name('certification-cells.json').read_text())['cells']
ENDPOINT = os.environ['CAVEMAN_MIDDLEWARE_ENDPOINT']


def digest(value):
    return hashlib.sha256((value if isinstance(value, str) else json.dumps(value, sort_keys=True, ensure_ascii=False, separators=(',', ':'))).encode()).hexdigest()


class CapturedRuntime(MiddlewareRuntime):
    def __init__(self, mode, report_callback=None, **kwargs):
        self.reports, self.plans, self.pages, self.receipts = [], [], [], []
        def report(value):
            self.reports.append(value)
            if report_callback:
                report_callback(value)
        super().__init__(endpoint='http://127.0.0.1:1' if mode == 'outage' else ENDPOINT, mode='off' if mode == 'off' else 'compress', deadline_ms=3000, on_report=report, **kwargs)
    def optimize(self, **options):
        result = super().optimize(**options)
        self.plans.append((options, result))
        return result
    def retrieve(self, scope, **args):
        page = super().retrieve(scope, **args)
        self.pages.append(page)
        return page
    def observe_background(self, receipt):
        self.receipts.append(receipt)
        return super().observe_background(receipt)


class NativeFixture(Provider):
    def __init__(self, protocol, *, structured=False, streaming=False):
        self.structured = structured
        super().__init__(protocol, pause='first' if streaming else None)
    def answer(self, body):
        text, call = super().answer(body)
        return json.dumps({'answer': FACT}) if self.structured and not call else text, call


def params(server, provider, **kwargs):
    return {'model': 'anthropic/claude-sonnet-4-20250514' if provider == 'anthropic' else 'openai/fixture-model',
            'api_key': 'local-fixture', 'api_base': server.url + '/v1', 'max_tokens': 100, 'max_retries': 0, **kwargs}


def source_text(body, protocol, call_id='read-1'):
    value = results(body, protocol).get(call_id)
    return ''.join(part.get('text', '') for part in value) if isinstance(value, list) else value


async def native_call(bridge, method, scope, kwargs):
    function = getattr(bridge, method)
    return await function(scope=scope, **kwargs) if method.startswith('a') else await asyncio.to_thread(function, scope=scope, **kwargs)


async def collect(test, stream, server, asynchronous, responses):
    test.assertTrue(callable(getattr(stream, '__aiter__' if asynchronous else '__iter__', None)))
    first = await asyncio.wait_for(anext(stream), 8) if asynchronous else await asyncio.wait_for(asyncio.to_thread(next, stream), 8)
    test.assertFalse(server.release.is_set(), 'Native first event must arrive before provider EOF')
    test.assertFalse(server.stream_done.is_set(), 'Provider must still be paused when the native event arrives')
    server.release.set()
    chunks = [first] + ([chunk async for chunk in stream] if asynchronous else await asyncio.to_thread(list, stream))
    if responses:
        events = [chunk.type for chunk in chunks]
        test.assertIn('response.completed', events)
        completed = next(chunk.response for chunk in chunks if chunk.type == 'response.completed')
        text = completed.output[0].content[0].text
    else:
        test.assertIsInstance(stream, litellm.CustomStreamWrapper)
        test.assertTrue(all(isinstance(chunk, litellm.ModelResponseStream) for chunk in chunks))
        events = [chunk.choices[0].finish_reason or ('text' if chunk.choices[0].delta.content else 'metadata') for chunk in chunks if chunk.choices]
        text = ''.join(chunk.choices[0].delta.content or '' for chunk in chunks if chunk.choices)
    test.assertEqual(text, FACT)
    return text, events, type(stream).__name__


async def settled(test, runtime):
    for _ in range(500):
        events = [receipt['event_kind'] for receipt in runtime.receipts]
        if events.count('completed') + events.count('failed') >= events.count('dispatch_intent'):
            return
        await asyncio.sleep(.01)
    test.fail('Native success/failure callbacks did not settle: ' + str([r['event_kind'] for r in runtime.receipts]))


def reporting(test, runtime, *, expected=None, allow_unscoped=False):
    if expected is not None:
        test.assertEqual(len(runtime.reports), expected, 'One report per owned dispatch or disabled public call')
    test.assertTrue(runtime.reports)
    attempt_ids = [report.attempt_id for report in runtime.reports if report.attempt_id is not None]
    test.assertEqual(len(set(attempt_ids)), len(attempt_ids))
    if not allow_unscoped:
        test.assertEqual(len(attempt_ids), len(runtime.reports))
    test.assertIs(runtime.last_report, runtime.reports[-1])
    test.assertNotIn(SOURCE, json.dumps([asdict(report) for report in runtime.reports]))
    test.assertEqual(sum(report.replacement_count for report in runtime.reports), sum(len(result.replacements) for _, result in runtime.plans))
    if runtime.mode == 'off':
        test.assertEqual(runtime.plans, []); test.assertEqual(runtime.receipts, [])
        test.assertTrue(all(report.status == 'disabled' for report in runtime.reports))
    return {'count': len(runtime.reports), 'statuses': sorted(report.status for report in runtime.reports), 'replacement_count': sum(report.replacement_count for report in runtime.reports), 'unique_attempt_ids': True, 'attempt_id_count': len(attempt_ids), 'calls_without_attempt_id': len(runtime.reports)-len(attempt_ids), 'source_content_absent': True,
            'scope': 'public_method_when_disabled' if runtime.mode == 'off' else 'owned_provider_dispatch'}


def record(test, runtime, server, *, native_type, value, source_calls, history_original=True, events=None, extra=None, report_calls=None, allow_unscoped_reports=False):
    test.assertEqual(server.errors, [])
    dispatch = [r for r in runtime.receipts if r['event_kind'] == 'dispatch_intent']
    completed = [r for r in runtime.receipts if r['event_kind'] == 'completed']
    failed = [r for r in runtime.receipts if r['event_kind'] == 'failed']
    test.assertEqual(len(dispatch), len(runtime.plans))
    test.assertEqual(len(completed) + len(failed), len(dispatch))
    return {'provider_calls': len(server.calls), 'source_executions': source_calls, 'source_sha256': digest(SOURCE), 'source_utf8_bytes': len(SOURCE.encode()),
            'recovery_requests': len(runtime.pages), 'replacements': sum(len(result.replacements) for _, result in runtime.plans),
            'native_type': native_type, 'native_value': value, 'original_history': history_original, 'native_events': events or [],
            'dispatch_receipts': len(dispatch), 'completed_receipts': len(completed), 'failed_receipts': len(failed),
            'native_reports': reporting(test, runtime, expected=report_calls, allow_unscoped=allow_unscoped_reports), **(extra or {})}


async def model_case(test, cell, mode):
    provider, streaming = cell['provider'], cell['streaming']
    responses, asynchronous = 'responses' in cell['method'], cell['execution'] == 'async'
    structured = cell['structured_output']
    wire_protocol = 'anthropic-messages' if provider == 'anthropic' else 'openai-responses' if responses else 'openai-chat'
    app_protocol = 'openai-responses' if responses else 'openai-chat'
    method = ('a' if asynchronous else '') + ('responses' if responses else 'completion')
    source_calls = []
    def read_logs():
        source_calls.append({}); return SOURCE
    source = read_logs()
    incoming = history(app_protocol)
    incoming[-1]['output' if responses else 'content'] = source
    test.assertEqual(source_text({'input' if responses else 'messages': incoming}, app_protocol), source)
    original = copy.deepcopy(incoming)
    with NativeFixture(wire_protocol, structured=structured, streaming=streaming) as server, CapturedRuntime(mode) as runtime:
        with CavemanLiteLLM(runtime=runtime) as bridge:
            options = params(server, provider, **{'input' if responses else 'messages': incoming}, stream=streaming)
            if structured:
                options['response_format'] = {'type': 'json_object'}
            output = await native_call(bridge, method, Scope('litellm-exact', digest(cell['id'] + mode)), options)
            if streaming:
                value, events, native_type = await collect(test, output, server, asynchronous, responses)
            else:
                value = output.output[0].content[0].text if responses else output.choices[0].message.content
                if structured:
                    value = json.loads(value); test.assertEqual(value, {'answer': FACT})
                else:
                    test.assertEqual(value, FACT)
                native_type, events = type(output).__name__, []
            test.assertEqual(incoming, original); test.assertEqual(source_calls, [{}]); test.assertEqual(len(server.calls), 1)
            test.assertEqual(source_text(server.calls[0]['body'], wire_protocol), SOURCE)
            test.assertEqual(runtime.pages, [])
            test.assertTrue(all(not result.replacements and options['binding'] is None for options, result in runtime.plans))
            await settled(test, runtime)
            outcome = record(test, runtime, server, native_type=native_type, value=value, source_calls=1, events=events, report_calls=1,
                             extra={'native_api': method, 'request_sha256': digest(server.calls[0]['body']), 'application_tool_result_supplied_to_native_model': True, 'first_event_before_provider_eof': streaming})
        test.assertNotIn(bridge, litellm.callbacks)
        return outcome


def emit(test, cell, assertion, observation):
    print('CAVEMAN_MIDDLEWARE_OBSERVATION ' + json.dumps({'cell_id': cell['id'], 'test_id': f'{TEST_FILE}::{type(test).__name__}.{test._testMethodName}', 'assertion': assertion, 'observation': observation}), flush=True)


async def certify_provider(test, provider):
    for cell in CELLS:
        if cell['provider'] != provider:
            continue
        modes = {}
        for mode in ('compress', 'off', 'outage'):
            try:
                if cell['method'] == 'router.retry_and_fallback':
                    from _router_cases import router_case
                    modes[mode] = await router_case(test, cell, mode)
                elif cell['method'].startswith('proxy.'):
                    from _proxy_cases import proxy_case
                    modes[mode] = await proxy_case(test, cell, mode)
                elif cell['method'] == 'asgi_composition':
                    from _asgi_case import asgi_case
                    modes[mode] = await asgi_case(test, cell, mode)
                else:
                    modes[mode] = await model_case(test, cell, mode)
            except BaseException as error:
                raise AssertionError(f"{cell['id']} ({mode}): {error}") from error
        if cell['recovery'] == 'model_only':
            test.assertEqual(modes['compress']['request_sha256'], modes['off']['request_sha256'])
            test.assertEqual(modes['compress']['request_sha256'], modes['outage']['request_sha256'])
        for assertion in ('native_application', 'real_tool_result', 'transformed_provider_request', 'omitted_fact_requested', 'host_executes_exact_recovery', 'native_result_history_events_and_call_count'):
            record = {'outcome': 'observed', **modes['compress']}
            if cell['recovery'] == 'model_only' and assertion in ('transformed_provider_request', 'omitted_fact_requested', 'host_executes_exact_recovery'):
                record.update(outcome='recovery_free', reason='structured_output' if cell['structured_output'] else 'model_only_no_executor')
            emit(test, cell, assertion, record)
        emit(test, cell, 'off_baseline', {'outcome': 'observed', **modes['off']})
        emit(test, cell, 'optimizer_unavailable', {'outcome': 'observed', **modes['outage']})
